core/client/fs/src/main/java/alluxio/client/file/dora/netty/NettyClientStateMachine.java

```diff
@@ -0,0 +1,705 @@
+    TERMINATED_EXCEPTIONALLY,
+    CLIENT_CANCEL,
+    CLIENT_CANCEL_DATA_RECEIVED,
+    ACQUIRING_CHANNEL,
+    CHANNEL_ACTIVE,
+    RECEIVED_DATA,
+    EXPECTING_EOF,
+    RECEIVED_EOF
+  }
+
+  enum Trigger {
+    START,
+    CHANNEL_AVAILABLE,
+    CHANNEL_UNAVAILABLE,
+    DATA_AVAILABLE,
+    HEART_BEAT,
+    TIMEOUT,
+    INTERRUPTED,
+    EOF,
+    OUTPUT_LENGTH_FULFILLED,
+    OUTPUT_LENGTH_NOT_FULFILLED,
+    OUTPUT_ERROR,
+    SERVER_ERROR,
+    SERVER_CANCEL,
+    CHANNEL_ERROR,
+    // this trigger is intentionally left unaccepted by any state
+    // so that it will cause an unhandled trigger exception from the state machine
+    UNKNOWN_PAYLOAD,
+    DATA_DISCARDED,
+  }
+
+  private static class Triggers {
+    final TriggerWithParameters1<IOException, Trigger> mChannelUnavailable;
+    final TriggerWithParameters1<ByteBuf, Trigger> mDataAvailable;
+    final TriggerWithParameters1<AlluxioStatusException, Trigger> mServerError;
+    final TriggerWithParameters1<Throwable, Trigger> mChannelError;
+    final TriggerWithParameters1<IOException, Trigger> mOutputError;
+    final TriggerWithParameters1<TimeoutException, Trigger> mTimeout;
+    final TriggerWithParameters1<InterruptedException, Trigger> mInterrupted;
+
+    public Triggers(StateMachineConfig<State, Trigger> config) {
+      mChannelUnavailable =
+          config.setTriggerParameters(Trigger.CHANNEL_UNAVAILABLE, IOException.class);
+      mDataAvailable =
+          config.setTriggerParameters(Trigger.DATA_AVAILABLE, ByteBuf.class);
+      mServerError =
+          config.setTriggerParameters(Trigger.SERVER_ERROR, AlluxioStatusException.class);
+      mChannelError =
+          config.setTriggerParameters(Trigger.CHANNEL_ERROR, Throwable.class);
+      mOutputError =
+          config.setTriggerParameters(Trigger.OUTPUT_ERROR, IOException.class);
+      mTimeout =
+          config.setTriggerParameters(Trigger.TIMEOUT, TimeoutException.class);
+      mInterrupted =
+          config.setTriggerParameters(Trigger.INTERRUPTED, InterruptedException.class);
+    }
+  }
+
+  /**
+   * Constructor.
+   *
+   * @param context
+   * @param address
+   * @param requestBuilder
+   * @param outChannel
+   */
+  public NettyClientStateMachine(FileSystemContext context, WorkerNetAddress address,
+      Protocol.ReadRequest.Builder requestBuilder, WritableByteChannel outChannel) {
+    mContext = context;
+    AlluxioConfiguration conf = context.getClusterConf();
+    mReadTimeoutMs = conf.getMs(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+    mMaxPacketsInFlight = conf.getInt(PropertyKey.USER_NETWORK_NETTY_READER_BUFFER_SIZE_PACKETS);
+    mAddress = address;
+    // clone the builder so that the initial values does not get overridden
+    mRequestBuilder = requestBuilder::clone;
+    mOffset = requestBuilder.getOffset();
+    mLength = (int) requestBuilder.getLength();
+    mOutputChannel = outChannel;
+
+    StateMachineConfig<State, Trigger> config = new StateMachineConfig<>();
```

Comment:
todo: this state machine config object can be reused. Does not need to be created on every read request.